iroha_p2p/src/peer.rs

```diff
@@ -0,0 +1,90 @@
+use std::borrow::BorrowMut;
+use std::cmp::Ordering;
+use std::future::Future;
+use std::io::{Error, ErrorKind};
+use std::net::SocketAddr;
+use std::ops::{Deref, DerefMut};
+use std::sync::Arc;
+
+use async_stream::stream;
+use futures::Stream;
+use tokio::io;
+use tokio::io::{AsyncReadExt, AsyncWriteExt, AsyncRead};
+use tokio::net::tcp::{OwnedReadHalf, OwnedWriteHalf};
+use tokio::net::TcpStream;
+use tokio::sync::Mutex;
+use tokio::time::Instant;
+
+use iroha_actor::{Actor, Context, Handler, MessageResponse};
+use iroha_crypto::{Hash, PublicKey};
+use iroha_logger::{debug, error, info, trace, warn};
+
+use crate::message::Message;
+use crate::peer_state::State;
+use iroha_data_model::prelude::PeerId;
+use crate::network::SendPeerMsg;
+
+pub struct Peer {
+    id: PeerId,
+    write: OwnedWriteHalf,
+    read: Option<OwnedReadHalf>,
+}
+
+impl Peer {
+    pub fn new(id: PeerId, stream: TcpStream) -> Self {
+        let (read, write) = stream.into_split();
+        Self { id, write, read: Some(read) }
+    }
+
+    // Message here is message which you send over tcp
+    fn stream(&mut self) -> impl Stream<Item = Result<Message, std::io::Error>> {
+        let mut read: OwnedReadHalf = self.read.take().unwrap();
+        stream! {
+            loop {
+                readable(&read).await?;
+                let message = read_message(&mut read).await;
+                yield message;
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Actor for Peer {
+    async fn on_start(&mut self, ctx: &mut Context<Self>) {
+        // Subscribe stream
+        ctx.notify_with(self.stream());
+    }
+}
+
+#[async_trait::async_trait]
+impl Handler<SendPeerMsg> for Peer {
+    type Result = ();
+
+    async fn handle(&mut self, msg: SendPeerMsg) {
+        let data = msg.msg;
+
+    }
+}
+
+pub async fn readable<T: AsRef<TcpStream>>(stream: T) -> io::Result<()> {
+    stream.as_ref().readable().await
+}
+
+pub async fn read_message(stream: &mut OwnedReadHalf) -> Result<Message, Error> {
+    let size = stream.read_u32().await? as usize;
+    if size > 0 {
+        let mut buf = vec![0u8; size];
```

Comment:
We might have mallicious peer which would send some big number here and will force our peer to crash.

This was also an issue for previous implementation. If you don't want to focus on it during this pr, could you create a jira issue for it?

Reply:
Also you can remove one tab and keep the context for readers. Like this:
```
let size = stream.read_u32().await? as usize;
if size <= 0 {
    return Err(io::Error::from(ErrorKind::InvalidInput))
}
let mut buf = vec![0u8; size];
...
```

Reply:
Yes, I know :)
Here will be some check against a valid constant.